src/gui/properties/piecesbar.cpp

```diff
@@ -0,0 +1,325 @@
+#endif
+        }
+
+    private:
+        QTextStream &m_stream;
+    };
+}
+
+void PiecesBar::showToolTip(const QHelpEvent *e)
+{
+    if (!m_torrent)
+        return;
+
+    QString toolTipText;
+    QTextStream stream(&toolTipText, QIODevice::WriteOnly);
+    bool showDetailedInformation = QApplication::keyboardModifiers().testFlag(Qt::ShiftModifier);
+    if (!showDetailedInformation) {
+        stream << simpleToolTipText();
+        stream << '\n' << tr("Hold Shift key for detailed information");
+    }
+    else {
+        stream << "<html><body>";
+        const int imagePos = e->pos().x() - borderWidth;
+        if ((imagePos >=0) && (imagePos < m_image.width())) {
+            PieceIndexToImagePos transform {m_torrent, m_image};
+            int pieceIndex = transform.pieceIndex(e->pos().x() - borderWidth);
+            QList<int> files {m_torrent->info().fileIndicesForPiece(pieceIndex)};
+
+            DetailedTooltipRenderer renderer(stream,
+                                             files.count() > 1 ? tr("Files in this piece:") : tr("Files in these pieces:"));
+
+            const bool isFileNameCorrectionNeeded = this->isFileNameCorrectionNeeded();
+            for (int f: files) {
+                QString filePath {m_torrent->info().filePath(f)};
+                if (!isFileNameCorrectionNeeded)
+                    filePath.replace(QLatin1String("/.unwanted"), QString());
+
+                renderer(Utils::Misc::friendlyUnit(m_torrent->info().fileSize(f)), filePath);
+            }
+        }
+        stream << "</body></html>";
+    }
+    stream.flush();
+
+    QToolTip::showText(e->globalPos(), toolTipText, this);
+}
+
+void PiecesBar::highlightFile(int imagePos)
+{
+    if (!m_torrent || (imagePos < 0) || (imagePos >= m_image.width()))
+        return;
+
+    PieceIndexToImagePos transform {m_torrent, m_image};
+
+    int pieceIndex = transform.pieceIndex(imagePos);
+    QList<int> fileIndices {m_torrent->info().fileIndicesForPiece(pieceIndex)};
+    if (fileIndices.count() == 1) {
+        QPair<int,int> fileBounds = m_torrent->info().piecesForFile(fileIndices.first());
+
+        QRect newHighlitedRegion {transform.imagePos(fileBounds.first), 0,
+                                  transform.imagePos(fileBounds.second) - transform.imagePos(fileBounds.first) + 1, m_image.height()};
+        if (newHighlitedRegion != m_highlitedRegion) {
+            m_highlitedRegion = newHighlitedRegion;
+            update();
+        }
+    }
+    else if (!m_highlitedRegion.isEmpty()) {
+        m_highlitedRegion = QRect();
+        update();
+    }
+}
+
+void PiecesBar::updatePieceColors()
+{
+    m_pieceColors = QVector<QRgb>(256);
+    for (int i = 0; i < 256; ++i) {
+        float ratio = (i / 255.0);
+        m_pieceColors[i] = mixTwoColors(backgroundColor().rgb(), m_pieceColor.rgb(), ratio);
+    }
+}
+
+bool PiecesBar::isFileNameCorrectionNeeded() const
+{
+    return false;
```

Comment:
The function is virtual, and the overrided one in `PieceAvailabilityBar` returns `true`.